### PR TITLE
Fix support for local functions in "Make Method Asynchronous"

### LINKED ISSUE #14133: Make Method Async applies to outermost method when run on Local Functions
**Version Used**: preview/25725

``` C#
using System.Threading.Tasks;

class C
{
    public void M1()
    {
        void M2()
        {
            await M3Async()$$;
        }
    }

    async Task<int> M3Async()
    {
        return 1;
    }
}
```

Invoke Make Method Async at the $$, and it will update M1 instead of M2.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #14133 

I also fixed a similar issue in the "Make Method Synchronous" code fix provider.

Tagging @dotnet/roslyn-ide  for review

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs [hunk 1]
```diff
@@ -599,5 +599,88 @@ class Program
...
+            await M3Async();
+        }
+    }
+
+    async Task<int> M3Async()
+    {
+        return 1;
+    }
+}",
+index: 1);
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: This is super subtle. So subtle I don't get it. All of the code in these two tests is the same right? Are there 2 code actions here that result in the same thing? What am I missing?

Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: I think the non-string arguments are usually indented to the normal level. Example: http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/AddBraces/AddBracesTests.cs,227

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs [hunk 2]
```diff
@@ -599,5 +599,88 @@ class Program
...
+    async Task<int> M3Async()
+    {
+        return 1;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeMethodAsynchronous)]
+        [WorkItem(14133, "https://github.com/dotnet/roslyn/issues/14133")]
+        public async Task AddAsyncInLocalFunction2()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: i don't see any difference between this and the previous test.

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: We're inconsistent with that.  I'm fine with this indentation.

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: Can we have a test with comment trivia before the local function.

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs [hunk 3]
```diff
@@ -19,14 +19,17 @@ internal class CSharpMakeMethodSynchronousCodeFixProvider : AbstractMakeMethodSy
 
         protected override bool IsMethodOrAnonymousFunction(SyntaxNode node)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: rename method to match current behavior.

File: src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs [hunk 4]
```diff
@@ -35,7 +35,9 @@ protected override string GetMakeAsyncVoidFunctionResource()
 
         protected override bool IsMethodOrAnonymousFunction(SyntaxNode node)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs]: rename method to match behavior.

File: src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs [hunk 5]
```diff
@@ -35,7 +35,9 @@ protected override string GetMakeAsyncVoidFunctionResource()
...
         {
-            return node.IsKind(SyntaxKind.MethodDeclaration) || node.IsAnyLambdaOrAnonymousMethod();
+            return node.IsKind(SyntaxKind.MethodDeclaration)
+                || node.IsAnyLambdaOrAnonymousMethod()
+                || node.IsKind(SyntaxKind.LocalFunctionStatement);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs]: duplicated code.  can we move to an extension?  (i actually think we have one that already does this somewhere).

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs [hunk 6]
```diff
@@ -51,7 +51,7 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
 
             // If it's a void returning method, offer to keep the void return type, or convert to 
             // a Task return type.
-            if (symbol?.MethodKind == MethodKind.Ordinary &&
+            if ((symbol?.MethodKind == MethodKind.Ordinary || symbol?.MethodKind == MethodKind.LocalFunction) &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: please don't mix ||s and &&s in expressions.  Just use nested ifs.  thanks.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs [hunk 7]
```diff
@@ -100,7 +100,7 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
 
-            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary &&
+            if ((methodSymbolOpt?.MethodKind == MethodKind.Ordinary || methodSymbolOpt?.MethodKind == MethodKind.LocalFunction) &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: same.

File: src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs [hunk 8]
```diff
@@ -49,7 +49,7 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
 
-            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary &&
+            if ((methodSymbolOpt?.MethodKind == MethodKind.Ordinary || methodSymbolOpt?.MethodKind == MethodKind.LocalFunction) &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs]: same.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: alternatively, extract method to do the || check, or introduce local.

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs (hunk 1 above)
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: @dpoeschl One of them changes the return type to `Task`, while the other leaves it as `void`.

Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: Can you name the test method names to match that description?

Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: 💭  Originally they were but then I changed them to match the less descriptive naming indicated by @CyrusNajmabadi in a different PR.

Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: I'd switch back to your original names, especially given how subtle the difference is between the two tests. There are plenty of examples of descriptive test names, way more than `TestWhatever#` examples. I'm sure @CyrusNajmabadi will agree. :smile:

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs (hunk 2 above)
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: :arrow_right: Added 18 tests. Fixed behavior for some and left two skipped per #18396.

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs (hunk 1 above)
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: ➡️ Renamed the tests.

File: src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs (hunk 4 above)
Comment (sharwell) [inline src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs]: :arrow_right: Renamed to `IsContainingFunction`

File: src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs (hunk 5 above)
Comment (sharwell) [inline src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs]: I couldn't find an existing extension for this. I always wasn't sure what to call it or where to put it if we did have it. For now I left it but if you have a more specific suggestion I can implement that. The functionality does seem specific to just these two operations, at least for now.

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs (hunk 3 above)
Comment (sharwell) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: :arrow_right: See previous case.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs (hunk 6 above)
Comment (sharwell) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: :arrow_right: Introduced local.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs (hunk 7 above)
Comment (sharwell) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: :arrow_right: Introduced local.

File: src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs (hunk 8 above)
Comment (sharwell) [inline src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs]: :arrow_right: Introduced local.

**Comment by sharwell** [comment]
@CyrusNajmabadi Can you check the updated handling of trivia (9857a18, 6e6001b) and the renamed methods (d8eac62)?

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs [hunk 9]
```diff
@@ -599,5 +599,180 @@ class Program
...
+
+    async Task<int> M3Async()
+    {
+        return 1;
+    }
+}",
+index: 1);
+        }
+
+        [Theory]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: i'm really not liking this Theory approach to these sorts of tests.  Reading the tests is now much harder.  And it makes basic stuff like { } in the test become distracting as {{ and }}   I don't see why this is actually better than just having two actual tests here.  One for the Task case and one for the void case.

A real purpose for these tests is for the text of the test to be self contained and self describing.  i.e. a single string that contains all the information necessary to know what's going on.  That's why we have the markup language.  There is no separate mechanism for specifying the caret position.  It's just in the text.  This 'Theory' approach fights against that, with negative results IMO.

I think theories can be good for very basic, non-replacing tests.  Where the theory/data contain all the information you need to know.  In that sense, it's a way of stamping out tests easily, but each test is still clear on what it does.  As that's not the case here 👎 on this part for me.

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs [hunk 10]
```diff
@@ -17,16 +17,19 @@ internal class CSharpMakeMethodSynchronousCodeFixProvider : AbstractMakeMethodSy
 
         public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(CS1998);
 
-        protected override bool IsMethodOrAnonymousFunction(SyntaxNode node)
+        protected override bool IsContainingFunction(SyntaxNode node)
         {
-            return node.IsKind(SyntaxKind.MethodDeclaration) || node.IsAnyLambdaOrAnonymousMethod();
+            return node.IsKind(SyntaxKind.MethodDeclaration)
+                || node.IsAnyLambdaOrAnonymousMethod()
+                || node.IsKind(SyntaxKind.LocalFunctionStatement);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: duplicate code.  we really should have an extension for this.

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs [hunk 11]
```diff
@@ -36,47 +39,66 @@ protected override SyntaxNode RemoveAsyncTokenAndFixReturnType(IMethodSymbol met
 
         private SyntaxNode FixMethod(IMethodSymbol methodSymbol, MethodDeclarationSyntax method, ITypeSymbol taskType, ITypeSymbol taskOfTType)
         {
-            var newReturnType = method.ReturnType;
+            var newReturnType = FixMethodReturnType(methodSymbol, method.ReturnType, taskType, taskOfTType);
+            SyntaxTokenList newModifiers = FixMethodModifiers(method.Modifiers, ref newReturnType);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: var

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs [hunk 12]
```diff
@@ -36,47 +39,66 @@ protected override SyntaxNode RemoveAsyncTokenAndFixReturnType(IMethodSymbol met
...
+            return method.WithReturnType(newReturnType).WithModifiers(newModifiers);
+        }
+
+        private SyntaxNode FixLocalFunction(IMethodSymbol methodSymbol, LocalFunctionStatementSyntax localFunction, ITypeSymbol taskType, ITypeSymbol taskOfTType)
+        {
+            var newReturnType = FixMethodReturnType(methodSymbol, localFunction.ReturnType, taskType, taskOfTType);
+            SyntaxTokenList newModifiers = FixMethodModifiers(localFunction.Modifiers, ref newReturnType);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: var

Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: how about taking in the return type, and returning it as a tuple with the modifiers.  Not a huge fan of 'ref' unless necessary for perf.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs [hunk 13]
```diff
@@ -100,8 +101,10 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
 
-            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary &&
-                !methodSymbolOpt.Name.EndsWith(AsyncSuffix))
+            bool isOrdinaryOrLocalFunction =
+                methodSymbolOpt?.MethodKind == MethodKind.Ordinary
+                || methodSymbolOpt?.MethodKind == MethodKind.LocalFunction;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: our general pattern is to keep && and || with the left side of the binary and to align the left and right when they wrap.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs [hunk 14]
```diff
@@ -51,8 +50,10 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
 
             // If it's a void returning method, offer to keep the void return type, or convert to 
             // a Task return type.
-            if (symbol?.MethodKind == MethodKind.Ordinary &&
-                symbol.ReturnsVoid)
+            bool isOrdinaryOrLocalFunction =
+                symbol?.MethodKind == MethodKind.Ordinary
+                || symbol?.MethodKind == MethodKind.LocalFunction;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: triplicated code.  please extract helper.

File: src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs (hunk 9 above)
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: I find the theory offers two benefits:

1. It's a form of duplicate code elimination, not unlike the other cases you requested
2. It helps ensure that you can't accidentally introduce slightly different behavior in two closely related cases. For example, this test revealed that changing a token from `Task` to `void` altered the behavior of the formatting engine somewhere. It's much harder to see these differences when the tests are separated.

Considering you get used to the `{{` and `}}` forms pretty quickly (it's no worse than `[|text|]` plus the code samples are small) and the editor highlights the replacements, I find this test form both clearer and more robust for handling a coupled test case like this.

@Pilchie what do you think?

Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: mailto:lrieger@precast-software.com

Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: I'm tentatively with @sharwell on this one, though I also find the `{{` to be tough to read, particularly in a GitHub diff.

Not sure if non-interpolated strings with single `{` and concats with ` + ` would be better though.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs]: 1. It's a form of duplicate code elimination

I don't see that as a virtue here.  For many test suites I could imagine parameterizing everything to the point that all you do is call a test method with something like "Test("public", "Task", "async", inMethod: true, isLocal: true, useSystem: true)" etc. etc.  Thse would just stamp out all the different variations and test the codepaths.  However, to me, the understandability of the test is gone.  I'm very much in favor of the test acting *completely* as the expected input and expected output.  In a way it acts like a spec would, saying: given a user providing precisely this, I expect to get precisely that.

Yes, lot so of the tests will have overlapping sections.  But it does not feel virtuous to be extracting that out.

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs (hunk 10 above)
Comment (sharwell) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: Implemented as `IsAsyncSupportingFunctionSyntax`

File: src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs (hunk 12 above)
Comment (sharwell) [inline src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs]: :memo: I didn't change the `ref` usage in this case. This usage was a side effect of the simple fact that `SyntaxTokenList` is a value type.

File: src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs (hunk 14 above)
Comment (sharwell) [inline src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs]: ➡️ Implemented as `IsOrdinaryMethodOrLocalFunction`

**Comment by sharwell** [comment]
@CyrusNajmabadi I rebased this PR, and corrected the use of `var` inside one of the existing commits.

The actually new commits relevant to fixing your comments are:

* 09f3547178070a55a98c8d2187799c5c799781b9: Extract helper `IsOrdinaryMethodOrLocalFunction`
* 1bb2cbd2fb38efbfd6df5764b8d417f6e4836f12: Extract helper `IsAsyncSupportingFunctionSyntax`

**Comment by sharwell** [comment]
@MattGertz for ask mode

## Ask Mode

**Customer scenario**

Use the Make Method Asynchronous or Make Method Synchronous refactoring operations inside of a local function.

**Bugs this fixes:**

#14133

**Workarounds, if any**

Perform the operation manually.

**Risk**

Low; code changes are restricted to this refactoring operation. Existing tests were unaltered and new tests for several cases were added.

**Performance impact**

Negligible. The algorithmic complexity did not change, and no additional allocations on the fast path (user takes no action), or on the existing path (action does not involve local functions).

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Failed to add tests for a new language feature to an existing refactoring operation.

**How was the bug found?**

Internal user reported.

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by MattGertz** [comment]
Approved pending test completions.

**Pull request merged**